Start a design document (#2)
nushell/nupm: docs/design/README.md

Static dependencies:
* Advantages: reproducible, does not rely on system files (no more missing random.so.2), higher performance (allows joint optimization of dependencies and project itself)
* Disadvatage: increased compile time, binary size, can easily end up with multiple versions of the same library (hello Nushell dependencies)

Dynamic dependencies are the opposite basically. Note that Nushell currently supports only static dependencies, but we might be able to add the "linking" feature at some point.

We might want `nupm`support both types of dependencies.

## Package repository 

Packages need to be stored somewhere. There should be one central "official" location (see https://github.com/NixOS/nixpkgs for inspiration).

Additionally, user should be able to add 3rd party repositories as well as install local and other packages (e.g., from the web, just pointing at URL),
as long as it has `METADATA_FILE` telling `nupm` what to do.

## API / CLI Interface

Nushell's module design conflates CLI interface with API -- they are the same. Not all of the below are of the same priority.

> **Note**  
> commands like `list`, `install`, `search`, `uninstall`, `update`, ..., i.e. should

> > - give short descriptions by default
> > - give long descriptions with `--long-description (-l)`

Reviewer: ```suggestion
> - print short descriptions by default
> - print long descriptions with `--long-description (-l)`
```

I already mentioned it in some thread, but since I can do it as a suggestion then let's make your life simpler :D

Reviewer 2: done in ad6b0cd8a06722ce05f90a2f67bf782f40c87010

- `nupm new [--script] [--module]`
    - create a new local package with template files ([`kubouch/nuun`])
- `nupm list`
    - list currently installed packages and if they're out of date
- `nupm install`
    - install package into the currently active overlay (can override which overlay to install to)
    - `--reinstall (-r)`: reinstall package if installed
    - `--update (-u)`: update local packages if outdated

>     - `-u` > `-r`

Reviewer: ```suggestion
    - Both `-u` & `-r` flags might be specified, but updating has higher priority than reinstalling
```

Reviewer 2: done in ad6b0cd8a06722ce05f90a2f67bf782f40c87010

>     - `--yes (-y)`: do not ask for user confirmation to use `nupm install` in scripts

Reviewer: ```suggestion
    - `--yes (-y)`: do not ask for user confirmation
```

I would just remove it, one can have different reasons to disable interactive prompts

Reviewer 2: done in ad6b0cd8a06722ce05f90a2f67bf782f40c87010

as said below, i've made "to use `nupm install` in scripts" an example, not the only reason

- `nupm add`
    - add a dependency to the current project
    - it is different from `nupm install`: this one adds the dependency to the MANIFEST_FILE, `nupm install` does not
- `nupm uninstall`
    - uninstall a package from a currently active overlay (can override which overlay to install to)
    - `--yes (-y)`: do not ask for user confirmation to use `nupm uninstall` in scripts
- `nupm update`
    - update all packages in a currently active overlay (can specify package and/or overlay name)

>     - can be used to self-update: `nupm update nupm`, `nupm update --self` or `nupm update --all` (would update everything)

Reviewer: ```suggestion
    - can be used to self-update: `nupm update nupm`, `nupm update --self` or `nupm update --all` (the last one would update each `nupm` package, including `nupm`)
```

Let's not leave space for inambiguities 😇

>     - `--yes (-y)`: do not ask for user confirmation to use `nupm uninstall` in scripts
> - `nupm update`
>     - update all packages in a currently active overlay (can specify package and/or overlay name)
>     - can be used to self-update: `nupm update nupm`, `nupm update --self` or `nupm update --all` (would update everything)
>     - `--yes (-y)`: do not ask for user confirmation to use `nupm update` in scripts

Reviewer: ```suggestion
    - `--yes (-y)`: do not ask for user confirmation
- `nupm update`
    - update all packages in a currently active overlay (can specify package and/or overlay name)
    - can be used to self-update: `nupm update nupm`, `nupm update --self` or `nupm update --all` (would update everything)
    - `--yes (-y)`: do not ask for user confirmation
```

The same case as above

Reviewer 2: done in done in ad6b0cd8a06722ce05f90a2f67bf782f40c87010

i've just made the "script" an example

- `nupm search`
    - search package repository (only supported ones by default)
    - `--unsupported (-u)`: would also list packages that are not supported in the user's system, e.g. due to OS incompatibilities

>     - `--all (-a)`: would list all packages

Reviewer: ```suggestion
```

As mentioned in some thread, let's just stick to one flag

Reviewer 2: done in ad6b0cd8a06722ce05f90a2f67bf782f40c87010

## Installation, bootstraping [[toc](#table-of-content)]

Requires these actions from the user (this should be kept as minimal as possible):
* Add `NUPM_HOME/bin` to PATH (install location for binary projects)
* Add `NUPM_HOME/modules` to NU_LIB_DIRS
* Add `NUPM_HOME/overlays` to NU_LIB_DIRS
* Make the `nupm` command available somehow (e.g., `use` inside `config.nu`)

> :warning: **WIP**  
> The disadvantage of this is that the default install location is not an overlay. We could make `nupm` itself an overlay that adds itself as a command, so that you can activate/deactivate it. We might need a few attempts to get to the right solution.

There are several approaches:
* bootstrap using shell script sourced from web (like `rustup`)

> * embedded inside Nushell's binary

Reviewer 2: as proposed by @kubouch, let's start simple here and then, as the project becomes more mature, we'll see to move it to main `nushell/nushell` :relieved:
